File: rn-expo/src/mapping/speedMapping.ts
export type SpeedToVolumeConfig = {
  minSpeedMps: number;
  maxSpeedMps: number;
  minVolume: number;
  maxVolume: number;
};

export function mapSpeedToVolume(speedMps: number, config: SpeedToVolumeConfig): number {
  const { minSpeedMps, maxSpeedMps, minVolume, maxVolume } = config;
  const clampedSpeed = Math.min(Math.max(speedMps, minSpeedMps), maxSpeedMps);
  const t = (clampedSpeed - minSpeedMps) / (maxSpeedMps - minSpeedMps);
  const volume = minVolume + t * (maxVolume - minVolume);
  return Math.min(Math.max(volume, 0), 1);
}

File: lib/core/mapping/speed_mapping.dart
class SpeedToVolumeMapping {
  SpeedToVolumeMapping({
    required this.minSpeedMps,
    required this.maxSpeedMps,
    required this.minVolume,
    required this.maxVolume,
  })  : assert(minSpeedMps >= 0),
        assert(maxSpeedMps > minSpeedMps),
        assert(minVolume >= 0 && minVolume <= 1),
        assert(maxVolume >= 0 && maxVolume <= 1),
        assert(maxVolume >= minVolume);

  final double minSpeedMps;
  final double maxSpeedMps;
  final double minVolume;
  final double maxVolume;

  double map(double speedMps) {
    final clampedSpeed = speedMps.clamp(minSpeedMps, maxSpeedMps);
    final t = (clampedSpeed - minSpeedMps) / (maxSpeedMps - minSpeedMps);
    return (minVolume + t * (maxVolume - minVolume)).clamp(0.0, 1.0);
  }
}
